Build failure due to `NullPointerException`, related to `@Parameter(ref = ...)` processing

### Describe the bug

When trying to build project with `./mvn clean verify` (etc), with certain endpoint definitions I get following stack trace (first part, full is longer):

```
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 3.53 s <<< FAILURE! - in io.stargate.web.GreetingResourceTest
[ERROR] io.stargate.web.GreetingResourceTest.testHelloEndpoint  Time elapsed: 0.006 s  <<< ERROR!
java.lang.RuntimeException: 
java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.smallrye.openapi.deployment.SmallRyeOpenApiProcessor#build threw an exception: java.lang.NullPointerException: Cannot invoke "org.eclipse.microprofile.openapi.models.parameters.Parameter$In.equals(Object)" because the return value of "org.eclipse.microprofile.openapi.models.parameters.Parameter.getIn()" is null
	at io.smallrye.openapi.runtime.scanner.spi.AnnotationScanner.isPathParameter(AnnotationScanner.java:944)
	at io.smallrye.openapi.runtime.scanner.spi.AnnotationScanner.lambda$getRequestBodyParameterClassType$7(AnnotationScanner.java:930)
	at java.base/java.util.stream.IntPipeline$10$1.accept(IntPipeline.java:392)
	at java.base/java.util.stream.IntPipeline$10$1.accept(IntPipeline.java:393)
	at java.base/java.util.stream.Streams$RangeIntSpliterator.tryAdvance(Streams.java:82)
	at java.base/java.util.stream.IntPipeline.forEachWithCancel(IntPipeline.java:163)
	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:527)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:513)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:647)
	at io.smallrye.openapi.runtime.scanner.spi.AnnotationScanner.getRequestBodyParameterClassType(AnnotationScanner.java:936)
	at io.smallrye.openapi.runtime.scanner.spi.AnnotationScanner.processRequestBody(AnnotationScanner.java:875)
	at io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.processResourceMethod(JaxRsAnnotationScanner.java:437)
	at io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.lambda$processResourceMethods$0(JaxRsAnnotationScanner.java:279)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1845)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
	at io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.processResourceMethods(JaxRsAnnotationScanner.java:277)
	at io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.processResourceClass(JaxRsAnnotationScanner.java:247)
	at io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.processResourceClasses(JaxRsAnnotationScanner.java:222)
	at io.smallrye.openapi.jaxrs.JaxRsAnnotationScanner.scan(JaxRsAnnotationScanner.java:154)
	at io.smallrye.openapi.runtime.scanner.OpenApiAnnotationScanner.scan(OpenApiAnnotationScanner.java:127)
	at io.quarkus.smallrye.openapi.deployment.SmallRyeOpenApiProcessor.generateAnnotationModel(SmallRyeOpenApiProcessor.java:847)
	at io.quarkus.smallrye.openapi.deployment.SmallRyeOpenApiProcessor.build(SmallRyeOpenApiProcessor.java:706)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:944)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

Issue seems to be triggered with endpoint definition like so:

```
@Operation(...)
@APIResponses(...)
public Response getRowWithWhere(
   // possibly some other parameters, but this is the problem one:
   @Parameter(name = "raw", ref = RestOpenApiConstants.Parameters.RAW)
   // Note: if this was added, problem avoided
   // @QueryParam("raw")
          final boolean raw

```

wherein parameter in question uses a (valid) reference but does not have matching `@QueryParam` (or similar).
In my particular case annotations are defined in an `interface`, implemented by actual sub-class, if that matters. Parameter definition reference is defined in "application" class which is found.

It also looks like failure only occurs when `quarkus-resteasy-reactive` is used as the dependency (earlier had trouble to reproduce with "classic").

### Expected behavior

Build should either work or indicate the underlying problem (inability to figure out type of parameter for OpenAPI?); but would not fail with NPE.

### Actual behavior

Build fails as a side effect of an internal `NullPointerException` being thrown.
See the stack trace in the description.

I will also try to isolate the failure case in a smaller project, if possible. The full project won't be possible to include due to its size (although theoretically maybe I could create a branch as this is for an OSS project, not closed).


### How to Reproduce?

Here's a project for reproduction

https://github.com/tatu-at-datastax/quarkus-demo

where simple `./mvnw clean verify` should show failure.

### Output of `uname -a` or `ver`

Darwin Kernel Version 20.6.0

### Output of `java -version`

openjdk version "17.0.2" 2022-01-18

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.10.3.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Maven 3.8.4

### Additional information

_No response_

--- a/bom/application/pom.xml
+++ b/bom/application/pom.xml
@@ -43,7 +43,7 @@
         <smallrye-config.version>2.11.1</smallrye-config.version>
         <smallrye-health.version>3.2.1</smallrye-health.version>
         <smallrye-metrics.version>3.0.5</smallrye-metrics.version>
-        <smallrye-open-api.version>2.1.23</smallrye-open-api.version>
+        <smallrye-open-api.version>2.2.0</smallrye-open-api.version>
         <smallrye-graphql.version>1.7.0</smallrye-graphql.version>
         <smallrye-opentracing.version>2.1.1</smallrye-opentracing.version>
         <smallrye-fault-tolerance.version>5.5.0</smallrye-fault-tolerance.version>

--- a/extensions/smallrye-openapi/deployment/src/main/java/io/quarkus/smallrye/openapi/deployment/SmallRyeOpenApiProcessor.java
+++ b/extensions/smallrye-openapi/deployment/src/main/java/io/quarkus/smallrye/openapi/deployment/SmallRyeOpenApiProcessor.java
@@ -535,10 +535,11 @@ public class SmallRyeOpenApiProcessor {
     }
 
     private Map<String, String> getClassNamesMethodReferences(OpenApiFilteredIndexViewBuildItem apiFilteredIndexViewBuildItem) {
+        FilteredIndexView filteredIndex = apiFilteredIndexViewBuildItem.getIndex();
         List<AnnotationInstance> openapiAnnotations = new ArrayList<>();
         Set<DotName> allOpenAPIEndpoints = getAllOpenAPIEndpoints();
         for (DotName dotName : allOpenAPIEndpoints) {
-            openapiAnnotations.addAll(apiFilteredIndexViewBuildItem.getIndex().getAnnotations(dotName));
+            openapiAnnotations.addAll(filteredIndex.getAnnotations(dotName));
         }
 
         Map<String, String> classNames = new HashMap<>();
@@ -546,33 +547,36 @@ public class SmallRyeOpenApiProcessor {
         for (AnnotationInstance ai : openapiAnnotations) {
             if (ai.target().kind().equals(AnnotationTarget.Kind.METHOD)) {
                 MethodInfo method = ai.target().asMethod();
-                if (Modifier.isInterface(method.declaringClass().flags())) {
-                    Collection<ClassInfo> allKnownImplementors = apiFilteredIndexViewBuildItem.getIndex()
-                            .getAllKnownImplementors(method.declaringClass().name());
-                    for (ClassInfo impl : allKnownImplementors) {
-                        MethodInfo implMethod = impl.method(method.name(), method.parameterTypes().toArray(new Type[] {}));
-                        if (implMethod != null) {
-                            String implRef = JandexUtil.createUniqueMethodReference(impl, method);
-                            classNames.put(implRef, impl.simpleName());
-                        }
-                    }
-                } else if (Modifier.isAbstract(method.declaringClass().flags())) {
-                    Collection<ClassInfo> allKnownSubclasses = apiFilteredIndexViewBuildItem.getIndex()
-                            .getAllKnownSubclasses(method.declaringClass().name());
-                    for (ClassInfo impl : allKnownSubclasses) {
-                        MethodInfo implMethod = impl.method(method.name(), method.parameterTypes().toArray(new Type[] {}));
-                        if (implMethod != null) {
-                            String implRef = JandexUtil.createUniqueMethodReference(impl, method);
-                            classNames.put(implRef, impl.simpleName());
-                        }
-                    }
+                ClassInfo declaringClass = method.declaringClass();
+                Type[] params = method.parameterTypes().toArray(new Type[] {});
+
+                if (Modifier.isInterface(declaringClass.flags())) {
+                    addMethodImplementationClassNames(method, params, filteredIndex
+                            .getAllKnownImplementors(declaringClass.name()), classNames);
+                } else if (Modifier.isAbstract(declaringClass.flags())) {
+                    addMethodImplementationClassNames(method, params, filteredIndex
+                            .getAllKnownSubclasses(declaringClass.name()), classNames);
                 } else {
-                    String ref = JandexUtil.createUniqueMethodReference(method.declaringClass(), method);
-                    classNames.put(ref, method.declaringClass().simpleName());
+                    String ref = JandexUtil.createUniqueMethodReference(declaringClass, method);
+                    classNames.put(ref, declaringClass.simpleName());
                 }
             }
         }
         return classNames;
+    }
+
+    void addMethodImplementationClassNames(MethodInfo method, Type[] params, Collection<ClassInfo> classes,
+            Map<String, String> classNames) {
+        for (ClassInfo impl : classes) {
+            String simpleClassName = impl.simpleName();
+            MethodInfo implMethod = impl.method(method.name(), params);
+
+            if (implMethod != null) {
+                classNames.put(JandexUtil.createUniqueMethodReference(impl, implMethod), simpleClassName);
+            }
+
+            classNames.put(JandexUtil.createUniqueMethodReference(impl, method), simpleClassName);
+        }
     }
 
     private boolean isValidOpenAPIMethodForAutoAdd(MethodInfo method, DotName securityRequirement) {

--- a/extensions/smallrye-openapi/deployment/src/test/java/io/quarkus/smallrye/openapi/test/jaxrs/AutoTagFetchableResource.java
+++ b/extensions/smallrye-openapi/deployment/src/test/java/io/quarkus/smallrye/openapi/test/jaxrs/AutoTagFetchableResource.java
@@ -1,0 +1,16 @@
+package io.quarkus.smallrye.openapi.test.jaxrs;
+
+import java.util.List;
+
+import javax.ws.rs.GET;
+
+public abstract class AutoTagFetchableResource<T> implements AbstractAutoTagResource<T> {
+
+    @GET
+    abstract List<T> getAll();
+
+    @Override
+    public T getById(long id) {
+        return null;
+    }
+}

--- a/extensions/smallrye-openapi/deployment/src/test/java/io/quarkus/smallrye/openapi/test/jaxrs/AutoTagResource.java
+++ b/extensions/smallrye-openapi/deployment/src/test/java/io/quarkus/smallrye/openapi/test/jaxrs/AutoTagResource.java
@@ -1,11 +1,20 @@
 package io.quarkus.smallrye.openapi.test.jaxrs;
 
+import java.util.List;
+
 import javax.ws.rs.Path;
 
-@Path("/address")
-public class AutoTagResource implements AbstractAutoTagResource<String> {
+@Path("/tagged")
+public class AutoTagResource extends AutoTagFetchableResource<String> {
+
     @Override
     public String getById(long id) {
         return "Disney Land, Gate " + id;
     }
+
+    @Override
+    public List<String> getAll() {
+        return null;
+    }
+
 }

--- a/extensions/smallrye-openapi/deployment/src/test/java/io/quarkus/smallrye/openapi/test/jaxrs/AutoTagTestCase.java
+++ b/extensions/smallrye-openapi/deployment/src/test/java/io/quarkus/smallrye/openapi/test/jaxrs/AutoTagTestCase.java
@@ -7,34 +7,27 @@ import org.junit.jupiter.api.extension.RegisterExtension;
 import io.quarkus.test.QuarkusUnitTest;
 import io.restassured.RestAssured;
 
-public class AutoTagTestCase {
+class AutoTagTestCase {
     @RegisterExtension
     static QuarkusUnitTest runner = new QuarkusUnitTest()
             .withApplicationRoot((jar) -> jar
-                    .addClasses(OpenApiResourceWithNoTag.class, AutoTagResource.class, AbstractAutoTagResource.class));
+                    .addClasses(OpenApiResourceWithNoTag.class, AutoTagResource.class, AutoTagFetchableResource.class,
+                            AbstractAutoTagResource.class));
 
     @Test
-    public void testAutoSecurityRequirement() {
+    void testTagInOpenApi() {
         RestAssured.given().header("Accept", "application/json")
-                .when().get("/q/openapi")
+                .when()
+                .get("/q/openapi")
                 .then()
-                .log().body()
-                .and()
-                .body("paths.'/resource/annotated'.get.tags", Matchers.hasItem("From Annotation"))
-                .and()
-                .body("paths.'/resource/auto'.get.tags", Matchers.hasItem("Open Api Resource With No Tag"))
-                .and()
-                .body("paths.'/resource/auto'.post.tags", Matchers.hasItem("Open Api Resource With No Tag"));
-
-    }
-
-    @Test
-    public void testTagInOpenApi() {
-        RestAssured.given().header("Accept", "application/json")
-                .when().get("/q/openapi")
-                .then()
+                .log().ifValidationFails()
+                .assertThat()
                 .statusCode(200)
-                .body(Matchers.containsString("Auto Tag Resource"));
+                .body("paths.'/tagged'.get.tags", Matchers.hasItem("Auto Tag Resource"))
+                .body("paths.'/tagged/{id}'.get.tags", Matchers.hasItem("Auto Tag Resource"))
+                .body("paths.'/resource/annotated'.get.tags", Matchers.hasItem("From Annotation"))
+                .body("paths.'/resource/auto'.get.tags", Matchers.hasItem("Open Api Resource With No Tag"))
+                .body("paths.'/resource/auto'.post.tags", Matchers.hasItem("Open Api Resource With No Tag"));
     }
 
 }

--- a/jakarta/rewrite.yml
+++ b/jakarta/rewrite.yml
@@ -557,7 +557,7 @@ recipeList:
       newValue: 4.0.0-RC1
   - org.openrewrite.maven.ChangePropertyValue:
       key: smallrye-open-api.version
-      newValue: 3.0.0-RC1
+      newValue: 3.0.0-RC3
   - org.openrewrite.maven.ChangePropertyValue:
       key: microprofile-opentracing-api.version
       newValue: 3.0

